File: java/log-levels/src/main/java/LogLevels.java
import java.util.regex.Pattern;

public class LogLevels {

	public static String message(String logLine) {
		return logLine.replaceAll("\\[[A-Z]*]:", "").trim();
	}

	public static String logLevel(String logLine) {
		final var pattern = Pattern.compile("\\[([A-Z]*)]:");

		final var matcher = pattern.matcher(logLine);

		if (matcher.find()) {
			return matcher.group(1).toLowerCase();
		}

		return "";
	}

	public static String reformat(String logLine) {
		return "%s (%s)".formatted(message(logLine), logLevel(logLine));
	}
}

File: fsharp/log-levels/LogLevels.fs
module LogLevels

open System.Text.RegularExpressions

let private LogRegex = Regex("\[(?<loglevel>[A-Z]+)\]:\s+(?<message>.*)")

let private parse (group: string) (logLine: string) =
    let m = logLine.Trim() |> LogRegex.Match
    m.Groups.[group].Value

let message (logLine: string) : string = logLine |> parse "message"

let logLevel (logLine: string) : string = (logLine |> parse "loglevel").ToLower()

let reformat (logLine: string) : string =
    $"{message logLine} ({logLevel logLine})"
